enhancements/ingress/gateway-api-with-cluster-ingress-operator.md

```diff
@@ -295,6 +288,133 @@ configures Envoy accordingly.  Configuring a load balancer and DNS enables
+
+```yaml
+apiVersion: maistra.io/v2
+kind: ServiceMeshControlPlane
+metadata:
+  name: gwapi
+  namespace: openshift-ingress
+spec:
+  profiles:
+  - gateway-controller
+  runtime:
+    components:
+      pilot:
+        container:
+          env:
+            PILOT_GATEWAY_API_CONTROLLER_NAME: openshift.io/gateway-controller
+  version: v2.4
+  techPreview:         # [1]
+    controlPlaneMode:
+      ClusterScoped
+```
+[1]: Specifying `techPreview.controlPlaneMode: ClusterScoped` is currently required even when specifying `spec.profiles: gateway-controller` due to a known issue in OSSM.
+
+This SMCP uses OSSM's built-in "gateway-controller" profile to enable Gateway
+API.  Specifying this profile is equivalent to specifying the following
+configuration on the SMCP:
+
+```yaml
+spec:
+  techPreview:
+    controlPlaneMode:
+      ClusterScoped
+    gatewayAPI:
+      enabled: true
+      controllerMode: true
+  addons:
+    grafana:
+      enabled: false
+    jaeger:
+      install: {}
+      name: jaeger
+    kiali:
+      enabled: false
+    prometheus:
+      enabled: false
+  gateways:
+    egress:
+      enabled: false
+    ingress:
+      enabled: true
+      ingress: true
+  policy:
+    type: Istiod
+  security:
+    manageNetworkPolicy: false
+  tracing:
+    type: None
+```
+
+The above configuration enables Gateway API with a cluster-scoped control-plane.
+This means that OSSM watches Gateway API resources in all namespaces.  The above
+configuration also enables ingress and egress proxies for service mesh; a
+namespace can subsequently be added to the mesh using ServiceMeshMember and
+ServiceMeshMemberRoll CRs following existing practices (see [Adding services to
+a service mesh](https://docs.openshift.com/container-platform/4.12/service_mesh/v2x/ossm-create-mesh.html) and [Managing users and profiles](https://docs.openshift.com/container-platform/4.12/service_mesh/v2x/ossm-profiles-users.html#ossm-control-plane-profiles_ossm-profiles-users) in the Service Mesh
+product documentation).
+
+#### New Controller to Manage Load Balancers for Gateway Listeners
+
+This enhancement adds a new controller to manage load balancers for Gateways.
+This controller watches the gateways resource, scoped to the "openshift-ingress"
+namespace.  For example, suppose a Gateway has the following definition (some
+extraneous details are abbreviated):
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: Gateway
+metadata:
+  name: example-gateway
+  namespace: openshift-ingress
+spec:
+  listeners:
+  - name: stage-http
+    hostname: "*.stage.example.com"
+    port: 80
+  - name: stage-https
+    hostname: "*.stage.example.com"
+    port: 443
+  - name: prod-https
+    hostname: "*.prod.example.com"
+    port: 443
+```
+
+In this configuration, the Gateway defines three listeners with two different
+TCP ports: 80 and 443.  For this Gateway, the Ingress Operator creates the
```

Comment:
I'm a little confused, maybe I'm missing something. Automated deployments currently do this, they create a Deployment and a Service. Is that what you are trying to describe? Also, if we did this, how would the Gateway attach/link to the service? You'd need to configure `spec.addresses` for a manual deployment right?

For reference, here's the service that Automatic Deployments create:
```
apiVersion: v1
kind: Service
metadata:
  creationTimestamp: "2023-02-06T19:43:37Z"
  finalizers:
  - service.kubernetes.io/load-balancer-cleanup
  labels:
    gateway.istio.io/managed: istio.io-gateway-controller
  name: gateway
  namespace: gwapi
  ownerReferences:
  - apiVersion: gateway.networking.k8s.io/v1alpha2
    kind: Gateway
    name: gateway
    uid: 54fc38b9-1882-4cea-816f-264286fe4d10
  resourceVersion: "132541"
  uid: 3d08f142-b252-4c77-831a-414a77e11650
spec:
  allocateLoadBalancerNodePorts: true
  clusterIP: 172.30.79.218
  clusterIPs:
  - 172.30.79.218
  externalTrafficPolicy: Cluster
  internalTrafficPolicy: Cluster
  ipFamilies:
  - IPv4
  ipFamilyPolicy: SingleStack
  ports:
  - name: status-port
    nodePort: 30616
    port: 15021
    protocol: TCP
    targetPort: 15021
  - name: http
    nodePort: 32720
    port: 80
    protocol: TCP
    targetPort: 80
  selector:
    istio.io/gateway-name: gateway
  sessionAffinity: None
  type: LoadBalancer
status:
  loadBalancer:
    ingress:
    - hostname: a3d08f142b2524c77831a414a77e1165-840945604.us-east-1.elb.amazonaws.com
```

Reply:
Thanks, that's very helpful!  I didn't realize the automated deployment feature also created the service.  However, that raises a bevy of questions:

* Is the service always configured with `type: LoadBalancer`?  
* Is the service always configured with `externalTrafficPolicy: Cluster`?  Can we specify `externalTrafficPolicy: Local` instead?  
* Can we configure the service to use an NLB on AWS, or configure timeouts, health checks, and so on? 
* Why does the service define a "status-port" port?  What does this port expose, and how is it secured?  
* Why doesn't your example have a port for HTTPS?  
* How does Istio name the ports?  
* How does Istio handle the case where two listeners specify the same port?  
* Does Istio tie the service's lifecycle to the gateway's?  (The owner reference suggests that it does.)  
* What is the significance of the `gateway.istio.io/managed: istio.io-gateway-controller` label?  

I've changed the text, but it might need further changes based on the answers to the above questions.

Reply:
Looks like Istio uses the listener name for the first listener that specifies the port: https://github.com/istio/istio/blob/c2b0f368d18ba54f216baa0d0ef6e023a1ee75fc/pilot/pkg/config/kube/gateway/deploymentcontroller.go#L376-L388

Is that the behavior that you've observed?

Reply:
> Is the service always configured with externalTrafficPolicy: Cluster? Can we specify externalTrafficPolicy: Local instead?

Doesn't look like it. Istio uses a template: https://github.com/istio/istio/blob/c2b0f368d18ba54f216baa0d0ef6e023a1ee75fc/pilot/pkg/config/kube/gateway/templates/service.yaml and there's no specification of `externalTrafficPolicy`, which mean it uses the default of `cluster`.

> Why doesn't your example have a port for HTTPS?

Ah yea sorry I just had a HTTP port 80 listener in my example, no HTTPS.

> Does Istio tie the service's lifecycle to the gateway's? (The owner reference suggests that it does.)

I believe so. It deletes the service when you delete the gateway, so I'd interpret that as they are tied to together.

> What is the significance of the gateway.istio.io/managed: istio.io-gateway-controller label?

Sounds like it's just used for filtering: https://github.com/istio/istio/blob/c2b0f368d18ba54f216baa0d0ef6e023a1ee75fc/pilot/pkg/config/kube/gateway/deploymentcontroller.go#L138 

> Looks like Istio uses the listener name for the first listener that specifies the port:

Correct, that's the behavior I see, just tested it. However, you made me realize something: the default ingress gateway (istio-ingressgateway) only specifies port 80 and 443, so you'd need to edit istio-ingressgateway in order to use another port (maybe that's obvious, but I hadn't thought about that until now). I confirmed Istio doesn't reconcile your changes to the istio-ingressgateway service.

Reply:
> Annotations and labels on the Gateway will be copied to the Service and Deployment. This allows configuring things such as Internal load balancers that read from these fields.

Just to be clear, this doesn't apply to the default `istio-ingressgateway`, only automatic deployments. This is probably pretty obvious, but I'm just learning that Istio creates `istio-ingressgateway` and doesn't do anything else to it.

Reply:
> the default ingress gateway (istio-ingressgateway) only specifies port 80 and 443, so you'd need to edit istio-ingressgateway in order to use another port (maybe that's obvious, but I hadn't thought about that until now).

Hm, I hadn't realized that either.  So the istio-ingressgateway service has the same ports irrespective of what listeners you specify in the Gateway?

Reply:
Right, the same 80 and 443 ports. Here is a fresh istio-ingressgateway service just for reference (no gateways or httproutes):
```
apiVersion: v1
kind: Service
metadata:
  annotations:
  creationTimestamp: "2023-02-07T23:12:21Z"
  finalizers:
  - service.kubernetes.io/load-balancer-cleanup
  labels:
    app: istio-ingressgateway
    app.kubernetes.io/component: istio-ingress
    app.kubernetes.io/instance: gwapi
    app.kubernetes.io/managed-by: maistra-istio-operator
    app.kubernetes.io/name: istio-ingress
    app.kubernetes.io/part-of: istio
    app.kubernetes.io/version: 2.3.1-1-1
    istio: ingressgateway
    istio.io/rev: istio-ingress
    maistra-version: 2.3.1
    maistra.io/owner: gwapi
    maistra.io/owner-name: istio-ingress
    release: istio
  name: istio-ingressgateway
  namespace: gwapi
  ownerReferences:
  - apiVersion: maistra.io/v2
    blockOwnerDeletion: true
    controller: true
    kind: ServiceMeshControlPlane
    name: istio-ingress
    uid: 314c0d5c-edfe-4a5a-89f2-0fe049f5b4e3
  resourceVersion: "218075"
  uid: 9051354f-b9d8-4230-be60-6be2aafb16e1
spec:
  allocateLoadBalancerNodePorts: true
  clusterIP: 172.30.171.40
  clusterIPs:
  - 172.30.171.40
  externalTrafficPolicy: Cluster
  internalTrafficPolicy: Cluster
  ipFamilies:
  - IPv4
  ipFamilyPolicy: SingleStack
  ports:
  - name: status-port
    nodePort: 30390
    port: 15021
    protocol: TCP
    targetPort: 15021
  - name: http2
    nodePort: 30449
    port: 80
    protocol: TCP
    targetPort: 8080
  - name: https
    nodePort: 31445
    port: 443
    protocol: TCP
    targetPort: 8443
  selector:
    app: istio-ingressgateway
    istio: ingressgateway
  sessionAffinity: None
  type: LoadBalancer
status:
  loadBalancer:
    ingress:
    - hostname: a9051354fb9d84230be606be2aafb16e-1375246106.us-east-1.elb.amazonaws.com
```